apps/mcp-ui-poc/client/src/lib/claude-client.ts

```diff
@@ -0,0 +1,515 @@
+        let errorMessage = "";
+        if (error && typeof error === "object") {
+          const err = error as any;
+          errorMessage = err.message || JSON.stringify(err);
+        } else {
+          errorMessage = String(error);
+        }
+
+        // Check if it's a token limit error (various formats)
+        const isTokenLimitError =
+          errorMessage.includes("prompt is too long") ||
+          errorMessage.includes("maximum") ||
+          (errorMessage.includes("token") && errorMessage.includes("200000"));
+
+        if (isTokenLimitError && retryCount < maxRetries) {
+          console.warn(
+            `⚠️ Token limit exceeded (attempt ${retryCount + 1}/${maxRetries + 1}). Retrying with guidance...`
+          );
+          console.warn(`Error details: ${errorMessage.substring(0, 200)}...`);
+          lastError = error as Error;
+          retryCount++;
+          continue;
+        }
+
+        // Not a token limit error or out of retries - throw it
+        throw error;
+      }
+    }
+
+    // If we exhausted retries, throw the last error with context
+    throw new Error(
+      `Failed after ${maxRetries + 1} attempts. Last error: ${lastError?.message || "Unknown error"}. ` +
+        `Suggestion: Try a more specific query with filters, date ranges, or smaller result sets.`
+    );
+  }
+
+  private async sendMessageAttempt(
+    message: string,
+    toolsEnabled: boolean,
+    retryAttempt: number
+  ) {
+    const uiResources: UIResource[] = [];
+    let textResponse = "";
+
+    // Create a temporary conversation for this message only (no history)
+    const currentConversation: Anthropic.MessageParam[] = [
+      {
+        role: "user",
+        content: message,
+      },
+    ];
+
+    // Loop to handle tool use (Claude may need multiple rounds)
+    let continueLoop = true;
+    while (continueLoop) {
+      // Build retry context if this is a retry attempt
+      const retryContext =
+        retryAttempt > 0
+          ? `
+
+🔄 RETRY ATTEMPT ${retryAttempt} - TOKEN LIMIT EXCEEDED ON PREVIOUS ATTEMPT
+
+The previous attempt exceeded the 200K token limit because tool responses were too large.
+
+CRITICAL: You MUST use more restrictive parameters:
+- Use pagination: limit to 5-10 items maximum
+- Add filters: date ranges, categories, status filters
+- Request specific fields only (avoid fetching all data)
+- Use "where" clauses to narrow results
+- Prefer summary/count endpoints over full list endpoints
+
+Examples of good retry parameters:
+- commerce__list_products with limit=5 and where="categories.id=\"cat123\""
+- commerce__read_orders with limit=10 and where="createdAt > \"2024-01-01\""
+- Request only essential fields, not full objects
+
+If you cannot fulfill the request with limited data, explain to the user that the dataset is too large and suggest:
+1. Adding filters (category, date range, status, etc.)
+2. Using pagination to view smaller batches
+3. Being more specific in their query
```

Comment:
@ByronDWall Hilariously, this is us explaining to the machine how limited our API is. Get creative, robot!